### PR TITLE
[RFC] Bootcamp Curriculum

### LINKED ISSUE #8: Limit the number of topics covered per SLU
# Limit the Number of Topics Covered per SLU

## Context

The number of topics covered by SLUs is wildly inconsistent

Also, we try to fit too much in too little time.

## Detailed Description

This would affect all SLUs and function as a guideline for curriculum changes.

According to the [Program Structure](https://github.com/LDSSA/wiki/wiki/Starters-Academy-program-structure-overview,-and-main-concepts), we have 20 minutes of teaching materials per SLU.

Also, we have 15 minutes to present each SLU.

Thus, limited time means we are forced to make trade-offs and keep the essential.

## Possible Implementation

I think we should limit the number of topics covered in each SLU to 5 (4 minutes per topic).

Since presentations last 15 mins, we would have 3 minutes to present each topic.

### PR DESCRIPTION
**Pull request by jtascensao**
Closes #8, #10, #11

# Bootcamp Curriculum

## Contributing Guidelines

If you haven't, go over the Contributing Guidelines.

> Put an `x` in the box if you did.

- [x] I have read the **CONTRIBUTING** document.

## Related Issue

#11

## Description

Please refer to the original issue (above).

## To-do

- [x] Structure
- [x] SLU01
- [x] SLU02
- [x] SLU03
- [x] SLU04
- [x] SLU05
- [x] SLU06
- [x] SLU07
- [x] SLU08
- [x] SLU09
- [x] SLU10
- [x] SLU11
- [x] SLU12
- [x] SLU13
- [x] SLU14
- [x] SLU15
- [x] SLU16
- [x] SLU17
- [x] SLU18
- [x] SLU19

### CODE REVIEW

**Comment by jtascensao** [comment]
[Direct link](https://github.com/LDSSA/curriculum-development/blob/bootcamp-curriculum/curriculum/01-bootcamp.md) to the document.

**Comment by jtascensao** [comment]
Summarizing changes and the main points for discussion (as well as their rational):

**The main purpose of this PR is to get the discussion going, so shoot!** :gun:

# What is changing from last year?

## What's goes off

- Removal of one of the SLUs on Pandas (there are still two of them), due to
    - Student feedback
    - The Prep course will cover some of it, even if it covers only Numpy many things translate to Pandas directly
    - The progression in the beginning is too slow and in the end is too fast, I know part of it is by design and desirable we need more balance (otherwise the end of the bootcamp is guerilla warfare)
    - The remaining SLUs on Pandas need to be adapted
- Removal of the SLU on Advanced Statistics, due to
    - Student feedback, again
    - Although important, most topics are too broad to be explored properly (pre-requisite)
    - Apart from probability distributions, I rarely (if ever) did hypothesis testing, etc., in Python
    - Remaining units on statistics might need to be adapted (though probably not)
- Removal of the SLU on Unsupervised Learning
    - Not used anywhere else throughout the Academy
    - Too broad to be explored in a single SLU
    - Not as relevant in practice
- Removal of the SLU on Algorithms
    - In a single SLU, we were introducing all foundation algorithms
    - Feedback from students was that
         - It was really, really hard to follow
         - They didn't build the intuition on how they work
         - They didn't really got when and how to use the models
         - Some got lost and most spent most of the first hackathon figuring it out
     - Not enough depth.

## What goes in

- Inclusion of a Notation & Definitions guide, for students and teachers (happy to take this on)
- The old Algorithms SLU is now split into three different SLUs on fundamental algorithms
    - Inclusion of an SLU on Support Vector Machines (SVM)
    - Inclusion of an SLU on Tree-Based Models
    - Inclusion of an SLU on K-Nearest Neighbors
    - I think that transmiting the intuition about these models, when and ow to use them is really important if we are to train entry-level data scientists (and also for the rest of the Academy)
- Inclusion of an SLU on Ethics & Fairness.

## What goes rebranded

- Renamed Regression to Linear Regression, narrowing the scope for effectiveness
- Renamed Classification to Logistic Regression, for the same reason
- Renamed Basic Stats to Basic Stats with Pandas, for explicitness
- Renamed Intermediate Stats to Covariance & Correlation, also for explicitness
- Renamed Real World Data to Feature Engineering, for the same reason
- Renamed Model Validaton to Model Validation & Overfitting, because overfitting.

# Next steps

* Main topics per SLU (three topics at most per SLU)
* Detailed proposal on the content covered per topic, to be discussed with the Owner of each SLU
* Move on to the rest of the curriculum (and come back to this for an integrated perspective).

Now, :gun: :gun: :gun:.

**Comment by PedroGFonseca** [comment]
Seem like a bunch of highly sensible changes indeed

**Comment by dominguezmaria** [comment]
Hello! Nice work!
Agree with everything in general.
Some particular comments.

**On removing one pandas SLU**
The removal makes sense because it really seemed that we were spending a lot of time with pandas and then rushing on other, more important things.

Maybe we can leave `split-apply-combine` out, because a) there won't be enough time to cover it properly with just 2 SLUs b) I don't remember any situation where one **must** use a group by during the bootcamp c) we can cover it more calmly in a specialisation (data wrangling or time series)

**On rebranding the Regression and Classification SLUs**
I'm totally for the scope narrowing.
But I'm not too sure about the rebranding, because of the "trickyness" of the logistic regression being used for classification. I'm afraid that calling the SLU just "Logistic Regression" may confuse the students and make their lives harder later (when we talk about classification and they don't find any SLU named classification).

So, my suggestion would be either:
- Call these SLUs: `Regression with Linear Regression`and `Classification with Logistic Regression` (I like this one better)
OR
- Keep their names unchanged (i.e, `Regression` and `Classification`), and explain clearly at the beginning of the SLU that we'll use specific algorithms to illustrate the Regression/Classification problems, but that there are other algorithms to approach these problems.

**On the next steps**
Go go go!!

**Comment by jtascensao** [comment]
@dominguezmaria on the Pandas side, what about merging, joining, and concatenating? Do you think they are more important? 

They could very reasonably go into Data Wrangling as well.

**Comment by dominguezmaria** [comment]
> @dominguezmaria on the Pandas side, what about merging, joining, and concatenating? Do you think they are more important?
> 
> They could very reasonably go into Data Wrangling as well.

We should check on last years' SLUs and first hackathon if we ever needed these.
If we don't, I'd be really happy about moving them into Data Wrangling.
If we do:
- concatenating is not too complicated, I guess
- we can simplify the merging and joining by just presenting the merge, since merge is more generic than join (everything that one does with a join can also be done with a merge).

**Comment by jtascensao** [comment]
> We should check on last years' SLUs and first hackathon if we ever needed these.
If we don't, I'd be really happy about moving them into Data Wrangling.
If we do:
> - concatenating is not too complicated, I guess
> -  we can simplify the merging and joining by just presenting the merge, since merge is more generic than join (everything that one does with a join can also be done with a merge).


I don't think we use it in [the first hackathon](https://github.com/LDSSA/batch2-hackathon1/tree/master/data).

I also don't think we need it for any of the remaining SLUs, as we want to keep them as much to the point as possible. I can double check, but even we use it somewhere I would remove it, for simplicity sake.

I would definitely move it to the Data Wrangling specialization, @dominguezmaria, if we all agree.

**Comment by jtascensao** [comment]
> Call these SLUs: Regression with Linear Regressionand Classification with Logistic Regression (I like this one better).

Done! :100:

**Comment by dominguezmaria** [comment]
> > We should check on last years' SLUs and first hackathon if we ever needed these.
> > If we don't, I'd be really happy about moving them into Data Wrangling.
> > If we do:
> > 
> > * concatenating is not too complicated, I guess
> > * we can simplify the merging and joining by just presenting the merge, since merge is more generic than join (everything that one does with a join can also be done with a merge).
> 
> I don't think we use it in [the first hackathon](https://github.com/LDSSA/batch2-hackathon1/tree/master/data).
> 
> I also don't think we need it for any of the remaining SLUs, as we want to keep them as much to the point as possible. I can double check, but even we use it somewhere I would remove it, for simplicity sake.
> 
> I would definitely move it to the Data Wrangling specialization, @dominguezmaria, if we all agree.

Just another note, I'm not sure if we use it in the Time Series don't use these in the Time Series specialisation. But if we do, we could also change the order of these two specialisations and that way the problem would be solved!

**Review by PedroGFonseca — APPROVED** [review]

**Comment by jtascensao** [comment]
Included the detailed proposal for SLU02 and SLU03.

Also, some other changes:

- Renamed Pandas 201 to Subsetting Data in Pandas (back to what it was before)
- Renamed Data Visualization to Visualization with Pandas and Matplotlib.

Please ignore the list of Topics Not Covered for now.

Duty calls, but I'll to close a couple more units later today. :crossed_fingers:

**Comment by jtascensao** [comment]
Pushed the detailed content until SLU06 (included), feel free to jump in there and comment!

**Comment by PedroGFonseca** [comment]
One important note: `Group-apply-combine` is a key concept imho.  However if the first specialisation is now wrangling, that can work fine

**Comment by jtascensao** [comment]
This PR is final. Removing the WIP status!

**Comment by manugarri** [comment]
so @jtascensao im checking slu 18, and its changed a bit from last year. instead of explaining how to choose a model it focuses on algorithmic specifics

or the cheatsheet is just a sheet detailing the different hyperparameters per model?

im saying because last year i gave examples on how to evaluate models (in terms of training time versus predictive performance)

**Comment by jtascensao** [comment]
@manugarri

> instead of explaining how to choose a model it focuses on algorithmic specifics

Good point. The way we thought about it, this SLU is very focused on hyperparameter optimization and what hyperparameters are.

The rational is having very specific SLUs, whenever possible.

Model selection is detailed in SLU09. It would be helpful to have your opinion on that one as well, I think that criteria such as training and prediction time are well raised and could be included there if you agree.

> or the cheatsheet is just a sheet detailing the different hyperparameters per model?

We received feedback from students on:

* They thought we were not giving them enough intuition on the models
* The same goes for how can they fine tune them.

In that sense, we think it's important to summarize:

* The main hyperparameters for each model
* And what's the intuition behind them and when to use them.

So these would be guidelines. We can go about them in a multitude of ways and I think this is a decision to be made alongside the owner of the SLU.

**Comment by jtascensao** [comment]
Also and given the above, @manugarri would you change the SLU in any way? If so, can you propose an alternative structure?

**Comment by manugarri** [comment]
> Model selection is detailed in SLU09 
So last year we explained nested cross validation in the Hyperparameter tuning (which makes sense since that is when its applied). I believe we could add it to this SLU

 Regarding other "real world" selection criterion (performance, explainability), it could fit in SLU09 (dunno if its too cramped though), or be kept in this one

> 
    The main hyperparameters for each model
    And what's the intuition behind them and when to use them.

The algorithms themselves are explained in other SLUs, wouldnt it make more sense to define each model's hyperparams on their respective SLU?

**Comment by jtascensao** [comment]
> Regarding other "real world" selection criterion (performance, explainability), it could fit in SLU09 (dunno if its too cramped though), or be kept in this one.

I think it could fit, especially given how useful it is in real life. I will try to propose something but it will be included, as you suggested. It's important stuff.

> The algorithms themselves are explained in other SLUs, wouldnt it make more sense to define each model's hyperparams on their respective SLU?

Could be, it's a sensible solution. I wonder if it would be the best choice given that we didn't introduce hyperparameters yet by then. 

In case we did it, what exactly could we cover in SLU19 alternatively? I'm not totally sold on addressing model selection in the hyperparameters unit.

**Comment by manugarri** [comment]
>Could be, it's a sensible solution. I wonder if it would be the best choice given that we didn't introduce hyperparameters yet by then.

hmm that is tricky indeed, better to leave them in slu19 then? an alternative is teach the first algorithm , then teach hyperparamenter search, then the rest of algorithms

**Comment by gaborsomogyi** [comment]
@jtascensao I read through the RFC, and while content-wise I couldn't find things to adjust, I gathered some minor inconsistencies, mainly between the overviews and the details:
- SLU01 and SLU03 have both an extra step in detail, so the numbers won't match - I'd suggest numbering the importing part as 0, then starting at 1. to match up with the overview numbers
- SLU04 - 3rd part in overview is TBD, while in details it's there (Outlier Detection) - probably an oversight, suggestion: adjust the overview
- SLU06 - overview and detail topics differ, suggestion: use the ones from detail
- SLU07 and SLU08 - 3rd topic on both - overview has an extra 'the', while details part omits it - suggestion: choose one to make it consistent
- SLU14 - 2nd topic differs between overview and details - suggestion: use the one from details to make them consistent
- SLU17 - 1st topic, 3rd item - probably a typo, did you mean Re_a_ssess and roll-back?

**Comment by dominguezmaria** [comment]
@jtascensao sorry to be late for the party, but just had to opportunity to review this now 😞 

Overall, I'm a bit concerned about the amount of content that we're including in each SLU, taking in consideration that we only have 15 minutes for the presentation.
I know that we're teaching based on intuition instead of using complicated mathematical explanations, but still, there are some cases where we're trying to explain in 5 minutes what I remember learning 50x more slowly in college... (Not kidding, I have proof!)

The topics that are worrying me **a lot** are:
* SLU04: skew, kurtosis, pdf and cdf
* SLU08: gradient descent (even if just intro)
* SLU09: different types of regularisation for linear models
* SLU12: kernel trick and kernels

(Then there are others that are worrying me a bit...)

Basically, the dilemma is that the curriculum (as it is) is very high quality and complete. So we don't feel like leaving anything behind. But then, the pace we're proposing here is insane and I think that most of the SLUs have too much content to be presented (and properly learned by the students) in 15 minutes. We're simply trying to "meter o Rossio na Rua da Betesga".

So, my suggestion is that we do one of these two:
a) cut down some of the content (and we should think carefully about what should be excluded)
b) still include everything but change the SLU format a bit 😱

<a href="https://imgflip.com/i/2zano1"></a>

I know this is disruptive... But bare with me a bit 😅 So, we could do the following:
* An SLU could have a new section called _Homework_ / _More on this topic_ / _Advanced but important stuff_ / _etc_, with the topics that we think are important but a bit too much to be included on the 15 min presentation + 45 min exercises. 
* This section would have some more content, that would be presented in BLU style, i.e, explained in detail in a notebook and not in the presentation, and exercises. The student should only focus on this content and exercises when the first part of the SLU is finished, which will (very likely) only happen after the bootcamp, **at home**.
* These exercises must be submitted through the grader for the SLU to be complete, i.e, they are **not optional**. They are just homework.
* The extra content/exercises **should not exist in all the SLUs**. In fact, we should try to avoid it: they should only be there when we feel that the content is too important not to be transmitted to the students, but we just don't have enough time to do it in 1 hour.

I think that the most important thing that we learned from the bootcamp last year is that the students felt it was too much, and that most of the times they were just too tired to absorb new concepts. And this curriculum (despite brilliant), feels even denser than last year's...

Now, 🔫🔫🔫 😛

**Comment by jtascensao** [comment]
@dominguezmaria you break my heart but you have a point. Can we create an issue so that we can discuss it there, given that the PR is closed?
